iwyu_ast_util.cc

```diff
@@ -225,6 +234,133 @@ static bool IsImplicitConversionCtor(const CXXConstructorDecl* ctor) {
+      // provision.
+      if (handled_.count(pair{lhs_record, rhs_record}))
+        return true;
+      handled_.insert(pair{lhs_record, rhs_record});
+
+      if (range_size(lhs_record->fields()) != range_size(rhs_record->fields()))
+        return false;
+      for (auto [lhs_field, rhs_field] :
+           zip_equal(lhs_record->fields(), rhs_record->fields())) {
+        if (lhs_field->getIdentifier() != rhs_field->getIdentifier())
+          return false;
+        // No check for alignment specifiers because if the fields differ only
+        // in them, it is probably just a bug.
+        // The rules are applied to the field types recursively.
+        // Field types are guaranteed to be complete.
+        if (!CouldBeCompatible(lhs_field->getType(), rhs_field->getType(), true,
+                               provided_types)) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  } else if (const auto* rhs_ptr = rhs->getAs<PointerType>()) {
+    if (const auto* lhs_ptr = lhs->getAs<PointerType>()) {
+      QualType rhs_pointee = rhs_ptr->getPointeeType();
+      QualType lhs_pointee = lhs_ptr->getPointeeType();
+      return CouldBeCompatible(lhs_pointee, rhs_pointee, false, provided_types);
+    }
+    return false;
+  } else if (const auto* rhs_array = ctx_.getAsArrayType(rhs)) {
+    if (const auto* lhs_array = ctx_.getAsArrayType(lhs)) {
+      if (const auto* rhs_const_array =
+              dyn_cast<ConstantArrayType>(rhs_array)) {
+        if (const auto* lhs_const_array =
+                dyn_cast<ConstantArrayType>(lhs_array)) {
+          if (lhs_const_array->getZExtSize() != rhs_const_array->getZExtSize())
+            return false;
+        }
+      }
+      // C language requires that an array element type should be complete,
+      // hence full_type_guaranteed = true.
+      return CouldBeCompatible(lhs_array->getElementType(),
+                               rhs_array->getElementType(), true,
+                               provided_types);
+    }
+    return false;
+  } else if (const auto* rhs_func = rhs->getAs<FunctionProtoType>()) {
+    if (const auto* lhs_func = lhs->getAs<FunctionProtoType>()) {
+      if (lhs_func->getNumParams() != rhs_func->getNumParams())
+        return false;
+      for (auto [lhs_param, rhs_param] :
+           zip_equal(lhs_func->param_types(), rhs_func->param_types())) {
+        // Parameter qualifications are ignored.
+        if (!CouldBeCompatible(lhs_param.getUnqualifiedType(),
+                               rhs_param.getUnqualifiedType(), false,
+                               provided_types)) {
+          return false;
+        }
+      }
+      return CouldBeCompatible(lhs_func->getReturnType(),
+                               rhs_func->getReturnType(), false,
+                               provided_types);
+    }
+    return false;
+  } else if (auto* rhs_atomic = rhs->getAs<AtomicType>()) {
```

Comment:
Could these be `const auto*` too, or do we use non-const members? Same question for enum types below.